### PR TITLE
Generic `NonZero` post-stabilization changes.

### PR DESCRIPTION
**Pull request by reitermarkus**
Tracking issue: https://github.com/rust-lang/rust/issues/120257

r? @dtolnay

### CODE REVIEW

**Comment by lnicola** [comment]
In rust-analyzer we'd prefer an `#[allow]`, because it needs to build on stable.

File: library/core/src/num/nonzero.rs [hunk 1]
```diff
@@ -490,19 +489,20 @@ macro_rules! nonzero_integer {
         ///
         /// [null pointer optimization]: crate::option#representation
         #[$stability]
+        #[deprecated(since = "TBD", note = "replaced by the generic `NonZero` type")]
```
Comment (dtolnay) [inline library/core/src/num/nonzero.rs:492]: I don't think a deprecation of the NonZero* type aliases has been approved by the standard library team. It would be better to land all the other changes that can be landed without this deprecation (such as docs) as soon as possible, and move this deprecation into a separate PR for a slower team-wide consensus.

**Review by dtolnay — CHANGES_REQUESTED** [review]

File: library/core/src/num/nonzero.rs (hunk 1 above)
Comment (RalfJung) [reply library/core/src/num/nonzero.rs:492]: I also feel like it may be good to wait a few releases before deprecating, so that crates with a MSRV policy of stable-N for small N will fix the deprecation instead of allowing the lint (and then forgetting to fix the deprecation later).

Comment (reitermarkus) [reply library/core/src/num/nonzero.rs:492]: > wait a few releases before deprecating

Isn't that what `deprecated_in_future` is meant for?

In any case I will move it into a separate PR.

**Comment by RalfJung** [comment]
Sorry, I won't be able to do further reviews here. I'm not a libs developer anyway.

File: library/core/src/num/nonzero.rs [hunk 2]
```diff
@@ -842,56 +845,55 @@ macro_rules! nonzero_integer_signedness_dependent_impls {
...
                 // and MIN/-1 is checked because `self` is an unsigned int.
                 unsafe { intrinsics::unchecked_div(self, other.get()) }
             }
         }
 
-        #[stable(feature = "nonzero_div_assign", since = "1.79.0")]
-        impl DivAssign<$Ty> for $Int {
-            /// This operation rounds towards zero,
-            /// truncating any fractional part of the exact result, and cannot panic.
+        #[stable(feature = "nonzero_div_assign", since = "CURRENT_RUSTC_VERSION")]
```
Comment (dtolnay) [inline library/core/src/num/nonzero.rs:861]: This impl was already stabilized in what's currently beta. The change on this line looks like a mistake during rebase.

```suggestion
        #[stable(feature = "nonzero_div_assign", since = "1.79.0")]
```

File: library/core/src/num/nonzero.rs [hunk 3]
```diff
@@ -842,56 +845,55 @@ macro_rules! nonzero_integer_signedness_dependent_impls {
...
+            fn rem(self, other: NonZero<$Int>) -> $Int {
+                // SAFETY: Remainder by zero is checked because `other` is non-zero,
                 // and MIN/-1 is checked because `self` is an unsigned int.
                 unsafe { intrinsics::unchecked_rem(self, other.get()) }
             }
         }
 
-        #[stable(feature = "nonzero_div_assign", since = "1.79.0")]
-        impl RemAssign<$Ty> for $Int {
+        #[stable(feature = "nonzero_div_assign", since = "CURRENT_RUSTC_VERSION")]
```
Comment (dtolnay) [inline library/core/src/num/nonzero.rs:884]:
```suggestion
        #[stable(feature = "nonzero_div_assign", since = "1.79.0")]
```

**Review by dtolnay — CHANGES_REQUESTED** [review]

**Comment by dtolnay** [comment]
LGTM with the 2 stability attributes fixed.

@bors delegate+
